File: skills/creator-monitoring-assistant/app/services/feishu_client.py
#!/usr/bin/env python3
"""飞书多维表格客户端。"""

from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Any, Dict, List, Optional

import requests

from app.config import get_settings


class FeishuBitableClient:
    """飞书多维表格 API 封装。"""

    def __init__(
        self,
        app_id: Optional[str] = None,
        app_secret: Optional[str] = None,
        app_token: Optional[str] = None,
        table_id: Optional[str] = None,
    ):
        settings = get_settings()
        self.settings = settings
        self.app_id = app_id or settings.feishu_app_id
        self.app_secret = app_secret or settings.feishu_app_secret
        self.app_token = app_token or settings.feishu_app_token
        self.table_id = table_id or settings.feishu_table_id
        self.base_url = "https://open.feishu.cn/open-apis"
        self._access_token: Optional[str] = None
        self._expires_at: float = 0

        if (not self.app_id or not self.app_secret) and Path.home().joinpath(".openclaw/openclaw.json").exists():
            config = json.loads(Path.home().joinpath(".openclaw/openclaw.json").read_text(encoding="utf-8"))
            self.app_id = self.app_id or config.get("channels", {}).get("feishu", {}).get("appId", "")
            self.app_secret = self.app_secret or config.get("channels", {}).get("feishu", {}).get("appSecret", "")

    def _get_access_token(self) -> str:
        if self._access_token and time.time() < self._expires_at:
            return self._access_token

        response = requests.post(
            f"{self.base_url}/auth/v3/tenant_access_token/internal",
            json={"app_id": self.app_id, "app_secret": self.app_secret},
            timeout=30,
        )
        result = response.json()
        if result.get("code") != 0:
            raise Exception(f"获取飞书 token 失败: {result.get('msg')}")

        self._access_token = result["tenant_access_token"]
        self._expires_at = time.time() + result.get("expire", 7200) - 300
        return self._access_token

    def _headers(self) -> Dict[str, str]:
        return {
            "Authorization": f"Bearer {self._get_access_token()}",
            "Content-Type": "application/json",
        }

    def _request(self, method: str, url: str, **kwargs) -> Dict[str, Any]:
        last_error: Optional[Exception] = None
        for attempt in range(3):
            try:
                response = requests.request(method, url, headers=self._headers(), timeout=30, **kwargs)
                result = response.json()
                if result.get("code") != 0:
                    raise Exception(result)
                return result.get("data", {})
            except Exception as exc:
                last_error = exc
                if attempt < 2:
                    wait_time = 2 ** attempt
                    time.sleep(wait_time)
        raise Exception(f"飞书请求失败: {last_error}")

    def list_fields(self) -> List[Dict[str, Any]]:
        data = self._request(
            "GET",
            f"{self.base_url}/bitable/v1/apps/{self.app_token}/tables/{self.table_id}/fields",
            params={"page_size": 500},
        )
        return data.get("items", [])

    def list_all_records(self) -> List[Dict[str, Any]]:
        records: List[Dict[str, Any]] = []
        page_token = None
        page_size = min(max(self.settings.feishu_read_page_size, 1), 500)
        while True:
            params = {"page_size": page_size}
            if page_token:
                params["page_token"] = page_token
            data = self._request(
                "GET",
                f"{self.base_url}/bitable/v1/apps/{self.app_token}/tables/{self.table_id}/records",
                params=params,
            )
            records.extend(data.get("items", []))
            if not data.get("has_more"):
                break
            page_token = data.get("page_token")
            if not page_token:
                break
        return records

    def batch_create_records(self, records: List[Dict[str, Any]]) -> Dict[str, Any]:
        return self._request(
            "POST",
            f"{self.base_url}/bitable/v1/apps/{self.app_token}/tables/{self.table_id}/records/batch_create",
            json={"records": records},
        )

    def batch_update_records(self, records: List[Dict[str, Any]]) -> Dict[str, Any]:
        return self._request(
            "POST",
            f"{self.base_url}/bitable/v1/apps/{self.app_token}/tables/{self.table_id}/records/batch_update",
            json={"records": records},
        )

    def update_record(self, record_id: str, fields: Dict[str, Any]) -> Dict[str, Any]:
        return self._request(
            "PUT",
            f"{self.base_url}/bitable/v1/apps/{self.app_token}/tables/{self.table_id}/records/{record_id}",
            json={"fields": fields},
        )

    def delete_record(self, record_id: str) -> Dict[str, Any]:
        return self._request(
            "DELETE",
            f"{self.base_url}/bitable/v1/apps/{self.app_token}/tables/{self.table_id}/records/{record_id}",
        )

File: skills/jimeng-video-generator/lib/feishu-client.js
const https = require('https');
const fs = require('fs');
const path = require('path');

let cachedAccessToken = null;
let tokenExpireTime = 0;

function loadOpenclawFeishuCredentials() {
  try {
    const openclawConfigPath = path.join(process.env.HOME || '', '.openclaw', 'openclaw.json');
    if (!fs.existsSync(openclawConfigPath)) {
      return {};
    }
    const raw = JSON.parse(fs.readFileSync(openclawConfigPath, 'utf8'));
    return raw.channels?.feishu || {};
  } catch (error) {
    return {};
  }
}

function getFeishuRequestTimeoutMs(config) {
  return Math.max(5000, Number(activeConfigForNetwork?.feishuRequestTimeoutMs || config?.feishuRequestTimeoutMs) || 45000);
}

function getFeishuDownloadTimeoutMs(config) {
  return Math.max(10000, Number(activeConfigForNetwork?.feishuDownloadTimeoutMs || config?.feishuDownloadTimeoutMs) || 120000);
}

let activeConfigForNetwork = null;

function requestJson(method, requestPath, token, body = null) {
  return new Promise((resolve, reject) => {
    const payload = body ? JSON.stringify(body) : null;
    const headers = {
      'Content-Type': 'application/json'
    };

    if (token) {
      headers.Authorization = `Bearer ${token}`;
    }
    if (method === 'GET') {
      headers['Cache-Control'] = 'no-cache, no-store, max-age=0';
      headers.Pragma = 'no-cache';
    }
    if (payload) {
      headers['Content-Length'] = Buffer.byteLength(payload);
    }

    const timeoutMs = getFeishuRequestTimeoutMs();
    const req = https.request({
      hostname: 'open.feishu.cn',
      path: requestPath,
      method,
      headers
    }, res => {
      let data = '';
      res.setEncoding('utf8');
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const json = JSON.parse(data);
          if (json.code === 0) {
            resolve(json.data);
            return;
          }
          if (json.code === 91403) {
            reject(new Error('飞书权限不足（91403 Forbidden）'));
            return;
          }
          reject(new Error(`飞书 API 失败 (${json.code}): ${json.msg || '未知错误'}`));
        } catch (error) {
          reject(new Error(`解析飞书响应失败: ${error.message}`));
        }
      });
    });

    req.setTimeout(timeoutMs, () => {
      req.destroy(new Error(`飞书请求超时 (${timeoutMs}ms): ${method} ${requestPath}`));
    });
    req.on('error', reject);
    if (payload) {
      req.write(payload);
    }
    req.end();
  });
}

function requestRawJson(method, requestPath, body = null) {
  return new Promise((resolve, reject) => {
    const payload = body ? JSON.stringify(body) : null;
    const headers = {
      'Content-Type': 'application/json'
    };

    if (payload) {
      headers['Content-Length'] = Buffer.byteLength(payload);
    }

    const timeoutMs = getFeishuRequestTimeoutMs();
    const req = https.request({
      hostname: 'open.feishu.cn',
      path: requestPath,
      method,
      headers
    }, res => {
      let data = '';
      res.setEncoding('utf8');
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          resolve(JSON.parse(data));
        } catch (error) {
          reject(new Error(`解析飞书响应失败: ${error.message}`));
        }
      });
    });

    req.setTimeout(timeoutMs, () => {
      req.destroy(new Error(`飞书请求超时 (${timeoutMs}ms): ${method} ${requestPath}`));
    });
    req.on('error', reject);
    if (payload) {
      req.write(payload);
    }
    req.end();
  });
}

async function getAccessToken(config) {
  if (cachedAccessToken && Date.now() < tokenExpireTime) {
    return cachedAccessToken;
  }

  const data = await requestRawJson(
    'POST',
    '/open-apis/auth/v3/tenant_access_token/internal',
    {
      app_id: config.appId,
      app_secret: config.appSecret
    }
  );

  if (data.code !== 0 || !data.tenant_access_token) {
    throw new Error(`获取飞书 Token 失败: ${data.msg || '未知错误'}`);
  }

  cachedAccessToken = data.tenant_access_token;
  tokenExpireTime = Date.now() + ((data.expire || 7200) - 60) * 1000;
  return cachedAccessToken;
}

async function listTableFields(config, token) {
  const data = await requestJson(
    'GET',
    `/open-apis/bitable/v1/apps/${config.appToken}/tables/${config.tableId}/fields?page_size=200&_ts=${Date.now()}`,
    token
  );
  return data.items || [];
}

async function createField(config, token, fieldSpec) {
  return requestJson(
    'POST',
    `/open-apis/bitable/v1/apps/${config.appToken}/tables/${config.tableId}/fields`,
    token,
    fieldSpec
  );
}

async function updateField(config, token, fieldId, fieldSpec) {
  return requestJson(
    'PUT',
    `/open-apis/bitable/v1/apps/${config.appToken}/tables/${config.tableId}/fields/${fieldId}`,
    token,
    fieldSpec
  );
}

async function updateRecord(config, token, recordId, fields) {
  return requestJson(
    'PUT',
    `/open-apis/bitable/v1/apps/${config.appToken}/tables/${config.tableId}/records/${recordId}`,
    token,
    { fields }
  );
}

async function getRecord(config, token, recordId) {
  const data = await requestJson(
    'GET',
    `/open-apis/bitable/v1/apps/${config.appToken}/tables/${config.tableId}/records/${recordId}?_ts=${Date.now()}`,
    token
  );
  return data.record || null;
}

async function listAllRecords(config, token) {
  let pageToken = null;
  let records = [];

  do {
    let requestPath = `/open-apis/bitable/v1/apps/${config.appToken}/tables/${config.tableId}/records?page_size=${config.pageSize || 100}&_ts=${Date.now()}`;
    if (pageToken) {
      requestPath += `&page_token=${encodeURIComponent(pageToken)}`;
    }
    if (config.viewId) {
      requestPath += `&view_id=${encodeURIComponent(config.viewId)}`;
    }

    const data = await requestJson('GET', requestPath, token);
    records = records.concat(data.items || []);
    pageToken = data.page_token || null;
  } while (pageToken);

  return records;
}

function detectMimeType(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  if (ext === '.mp4') return 'video/mp4';
  if (ext === '.mov') return 'video/quicktime';
  if (ext === '.m4v') return 'video/x-m4v';
  if (ext === '.avi') return 'video/x-msvideo';
  if (ext === '.jpg' || ext === '.jpeg') return 'image/jpeg';
  if (ext === '.png') return 'image/png';
  if (ext === '.webp') return 'image/webp';
  if (ext === '.gif') return 'image/gif';
  return 'application/octet-stream';
}

async function uploadFileToFeishu(config, token, filePath) {
  const fileBuffer = fs.readFileSync(filePath);
  const filename = path.basename(filePath);
  const fileSize = fileBuffer.length;
  const mimeType = detectMimeType(filePath);

  const parentTypes = Array.isArray(config.uploadParentTypes) && config.uploadParentTypes.length > 0
    ? config.uploadParentTypes
    : ['bitable_file', 'bitable_image'];

  let lastError = null;

  for (const parentType of parentTypes) {
    try {
      const boundary = '----OpenClawJimeng' + Math.random().toString(16).slice(2);
      const CRLF = '\r\n';
      const parts = [];

      parts.push('--' + boundary + CRLF);
      parts.push('Content-Disposition: form-data; name="file_name"' + CRLF + CRLF);
      parts.push(filename + CRLF);

      parts.push('--' + boundary + CRLF);
      parts.push('Content-Disposition: form-data; name="parent_type"' + CRLF + CRLF);
      parts.push(parentType + CRLF);

      parts.push('--' + boundary + CRLF);
      parts.push('Content-Disposition: form-data; name="parent_node"' + CRLF + CRLF);
      parts.push(config.appToken + CRLF);

      parts.push('--' + boundary + CRLF);
      parts.push('Content-Disposition: form-data; name="size"' + CRLF + CRLF);
      parts.push(String(fileSize) + CRLF);

      parts.push('--' + boundary + CRLF);
      parts.push(`Content-Disposition: form-data; name="file"; filename="${filename}"` + CRLF);
      parts.push(`Content-Type: ${mimeType}` + CRLF + CRLF);

      const headerBuffer = Buffer.from(parts.join(''));
      const footerBuffer = Buffer.from(CRLF + '--' + boundary + '--' + CRLF);
      const payload = Buffer.concat([headerBuffer, fileBuffer, footerBuffer]);

      const data = await new Promise((resolve, reject) => {
        const req = https.request({
          hostname: 'open.feishu.cn',
          path: '/open-apis/drive/v1/medias/upload_all',
          method: 'POST',
          headers: {
            Authorization: `Bearer ${token}`,
            'Content-Type': `multipart/form-data; boundary=${boundary}`
          }
        }, res => {
          let raw = '';
          res.setEncoding('utf8');
          res.on('data', chunk => raw += chunk);
          res.on('end', () => {
            try {
              const json = JSON.parse(raw);
              if (json.code === 0) {
                resolve({
                  fileToken: json.data.file_token,
                  parentType
                });
                return;
              }
              reject(new Error(`上传失败 (${parentType}): ${json.msg || '未知错误'}`));
            } catch (error) {
              reject(new Error(`解析上传响应失败 (${parentType}): ${error.message}`));
            }
          });
        });

        req.on('error', reject);
        req.write(payload);
        req.end();
      });

      return data;
    } catch (error) {
      lastError = error;
    }
  }

  throw lastError || new Error('上传文件到飞书失败');
}

async function downloadFile(token, fileToken, outputPath) {
  const timeoutMs = getFeishuDownloadTimeoutMs();
  return new Promise((resolve, reject) => {
    const file = fs.createWriteStream(outputPath);
    const req = https.request({
      hostname: 'open.feishu.cn',
      path: `/open-apis/drive/v1/medias/${fileToken}/download`,
      method: 'GET',
      headers: {
        Authorization: `Bearer ${token}`
      }
    }, res => {
      if (res.statusCode !== 200) {
        let errorData = '';
        res.setEncoding('utf8');
        res.on('data', chunk => errorData += chunk);
        res.on('end', () => {
          fs.unlink(outputPath, () => {
            reject(new Error(`下载失败 (${res.statusCode}): ${errorData || '空响应'}`));
          });
        });
        return;
      }

      res.pipe(file);
      file.on('finish', () => {
        file.close(() => resolve(outputPath));
      });
    });

    req.setTimeout(timeoutMs, () => {
      req.destroy(new Error(`飞书文件下载超时 (${timeoutMs}ms): ${fileToken}`));
    });
    req.on('error', error => {
      file.close(() => {
        fs.unlink(outputPath, () => reject(error));
      });
    });
    req.end();
  });
}

function setActiveConfigForNetwork(config) {
  activeConfigForNetwork = config;
}

module.exports = {
  loadOpenclawFeishuCredentials,
  getFeishuRequestTimeoutMs,
  getFeishuDownloadTimeoutMs,
  requestJson,
  requestRawJson,
  getAccessToken,
  listTableFields,
  createField,
  updateField,
  updateRecord,
  getRecord,
  listAllRecords,
  detectMimeType,
  uploadFileToFeishu,
  downloadFile,
  setActiveConfigForNetwork
};
